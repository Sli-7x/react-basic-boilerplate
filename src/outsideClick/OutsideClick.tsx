import * as React from 'react';

export interface IOutsideClick {
  onOutside: () => void;
  children: any;
  wrapperRef?: any;
  onBlur?: () => void;
}

class OutsideClick extends React.Component<IOutsideClick, {}> {
  private wrapperRef: any;

  constructor(props: IOutsideClick) {
    super(props);

    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  /**
   * Set the wrapper ref
   */
  setWrapperRef(node: any) {
    this.wrapperRef = node;
  }

  /**
   * Alert if clicked on outside of element
   */
  handleClickOutside(event: Event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.props.onOutside();
    }
  }

  render() {
    return (
      <div ref={this.setWrapperRef} style={{ width: '100%', height: '100%' }} onBlur={this.props.onBlur}>
        {this.props.children}
      </div>
    );
  }
}

export default OutsideClick;
