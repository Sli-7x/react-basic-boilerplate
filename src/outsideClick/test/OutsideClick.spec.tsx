import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { OutsideClick, IOutsideClick } from '../OutsideClick';
import { shallow, mount } from 'enzyme';

describe('OutsideClick', () => {
  const onOutside = jest.fn();
  const props: IOutsideClick = {
    onOutside,
  };

  it('should render correctly', () => {
    const tree = renderer.create(<OutsideClick {...props}>Test</OutsideClick>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('componentWillUnmount()', () => {
    it('should call document.removeEventListener', () => {
      document.removeEventListener = jest.fn();
      const wrapper = shallow(<OutsideClick {...props} />);
      const instance = wrapper.instance() as OutsideClick;
      instance.componentWillUnmount();
      expect(document.removeEventListener).toHaveBeenCalled();
    });
  });

  describe('handleClickOutside()', () => {
    it('should not call onOutside()', () => {
      const wrapper = shallow(<OutsideClick {...props} />);
      const instance = wrapper.instance() as OutsideClick;
      const event = new Event('');
      instance.handleClickOutside(event);
      expect(onOutside).not.toHaveBeenCalled();
    });

    it('should call onOutside()', () => {
      const wrapper = shallow(<OutsideClick {...props} />);
      const instance = wrapper.instance() as OutsideClick;
      const w = { contains: () => {} };
      instance.setWrapperRef(w);
      const event = new Event('');
      instance.handleClickOutside(event);
      expect(onOutside).toHaveBeenCalled();
    });
  });
});
