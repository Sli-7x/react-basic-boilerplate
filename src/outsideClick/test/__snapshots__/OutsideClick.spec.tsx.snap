// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutsideClick should render correctly 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  Test
</div>
`;
